File: Ejercicios/JavaScript/Clases/Coche.js
class Coche {
  constructor(marca, modelo, anio) {
    if (marca === undefined && modelo === undefined && anio === undefined) {
      this.marca = "Genérico";
      this.modelo = "Base";
      this.anio = 2000;
    } else if (marca instanceof Coche) {
      this.marca = marca.marca;
      this.modelo = marca.modelo;
      this.anio = marca.anio;
    } else if (anio === undefined) {
      this.marca = marca;
      this.modelo = modelo;
      this.anio = 2020;
    } else {
      this.marca = marca;
      this.modelo = modelo;
      this.anio = anio;
    }
  }
}

export default Coche;

File: Ejercicios/GO/Clases/Coche.go
package main

type Coche struct {
	Marca  string
	Modelo string
	Anio   int
}

func NewCoche() Coche {
	return Coche{Marca: "Genérico", Modelo: "Base", Anio: 2000}
}

func NewCocheConDatos(marca, modelo string, anio int) Coche {
	return Coche{Marca: marca, Modelo: modelo, Anio: anio}
}

func NewCocheMarcaModelo(marca, modelo string) Coche {
	return Coche{Marca: marca, Modelo: modelo, Anio: 2020}
}

func NewCocheCopy(c Coche) Coche {
	return Coche{Marca: c.Marca, Modelo: c.Modelo, Anio: c.Anio}
}
